includes/class-amp-theme-support.php

```diff
@@ -1934,17 +1934,31 @@ public static function prepare_response( $response, $args = [] ) {
 			);
 		}
 
-		// Abort if an expected template was not rendered, in that template actions didn't fire and response type is not HTML.
-		$did_template_action = (
+		// Abort if response type is not HTML.
+		if ( Attribute::TYPE_HTML !== substr( AMP_HTTP::get_response_content_type(), 0, 9 ) ) {
+			return $response;
+		}
+
+		// Abort if an expected template action didn't fire or if the HTML tag does not have the AMP attribute.
+		if ( ! (
 			did_action( 'wp_head' )
 			||
 			did_action( 'wp_footer' )
 			||
 			did_action( 'amp_post_template_head' )
 			||
 			did_action( 'amp_post_template_footer' )
-		);
-		if ( ! $did_template_action || Attribute::TYPE_HTML !== substr( AMP_HTTP::get_response_content_type(), 0, 9 ) ) {
+			||
+			preg_match(
+				sprintf(
+					'#^(?:<!.*?>|\s+)*+<html[^>]*?\s(%1$s|%2$s|%3$s)(?:=".*?")?.*?(?>>.*?<)head[^>]*?>#is',
```

Comment:
This regex is slow for hits (257 steps on my test content), and leads to catastrophic backtracking for misses (if <head> is not found).

As I stated for a previous check, I don't see why we would need to check for the `<head>` here (which is the performance problem). If we detect AMP, we should process, and processing means we'll normalize the document structure anyway (which adds a <head> if needed).

This regex is much faster and should do the job:
```suggestion
					'#^(?:<!.*?>|\s+)*+<html\s+[^>]*?\s(%1$s|%2$s|%3$s)(\s|=|>)#is',
```

For the same content, this now uses 58 steps on a hit and 36 steps on a miss.

Reply:
The idea behind checking for `<head>` was just to have extra assurance that response is a full HTML document. However, in the previous check I believe the context was different: we weren't looking for an `amp` attribute. Now the the check for an HTML document which is suitable for conversion into AMP is done by checking to see if one of the head/footer actions were fired. If one of those don't fire (which will be rare), then this fallback scenario comes into play now where we check of the HTML document is _explicitly_ marked as an AMP page. So yes, I think with this new constraint we don't don't need to check for `<head>`.

Reply:
However, the regex is not quite right. It needs to be modified as follows:

```diff
- 					'#^(?:<!.*?>|\s+)*+<html\s+[^>]*?\s(%1$s|%2$s|%3$s)(\s|=|>)#is',
+ 					'#^(?:<!.*?>|\s+)*+<html(?=\s)[^>]*?\s(%1$s|%2$s|%3$s)(\s|=|>)#is',
```

So:

```suggestion
					'#^(?:<!.*?>|\s+)*+<html(?=\s)[^>]*?\s(%1$s|%2$s|%3$s)(\s|=|>)#is',
```

As otherwise it requires at least two whitespace chars to be between `<html` and `amp`.